### PR TITLE
Allow TuplePattern to have null fields and match any tuple

### PR DESCRIPTION
**Pull request by mbrookhart**
Fixes issue raised in https://discuss.tvm.apache.org/t/tuplepattern-for-any-number-of-inputs-to-the-tuple/8926

I'm not sure I love the fix, the specialization of paritioning to support ops that are technically outside of the pattern makes me queasy. Any thoughts to make it better?

@masahi

### CODE REVIEW

File: src/relay/ir/dataflow_matcher.cc [hunk 1]
```diff
@@ -400,12 +400,16 @@ bool DFPatternMatcher::VisitDFPattern_(const TupleGetItemPatternNode* op, const
...
-    if (op->fields.size() == tuple_node->fields.size()) {
-      matches = true;
-      size_t i = 0;
-      while (matches && i < op->fields.size()) {
-        matches &= VisitDFPattern(op->fields[i], tuple_node->fields[i]);
-        ++i;
+    matches = true;
+    if (op->fields.defined()) {
+      if (op->fields.size() == tuple_node->fields.size()) {
+        size_t i = 0;
```
Comment (masahi) [inline src/relay/ir/dataflow_matcher.cc:406]: `if (op->fields.defined() && op->fields.size() == tuple_node->fields.size()) {`

**Review by masahi — APPROVED** [review]
I think it's ok, but I'd like to see `is_tuple([wildcard()+])` supported in the future.

**Comment by slyubomirsky** [comment]
Q: Can you also have this for function calls? (Matching a call with any number of arguments.)

**Comment by mbrookhart** [comment]
Hmm, probably, @masahi and i have been talking about cleaner ways to do this for function/call/etc. Let me do some experiments and try to get back to you next week. I'm going to close this to work on a more complete solution.

**Pull request closed without merging**